Added _repr_html_ for fonts

ambi7 opened the pull request:
## PR Summary

Resolves [#18039](https://github.com/matplotlib/matplotlib/issues/18039)

Invoking in IPython
```python
import matplotlib.font_manager
from IPython.core.display import HTML

font = matplotlib.font_manager.fontManager.ttflist[0]       # any FontEntry

HTML(font._repr_html_())
```

## PR Checklist

<!-- Please mark any checkboxes that do not apply to this PR as [N/A]. -->
**Tests and Styling**
- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [N/A] New features are documented, with examples if plot related.
- [N/A] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [N/A] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [N/A] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Do not create the PR out of main, but out of a separate branch.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

anntzer reviewed (changes requested):
As noted in #18039 I don't think this implementation is satisfactory; for example STIXGeneral is only provided by matplotlib on my computer and therefore trying to display in with this mechanism (which uses the browser's OS integration to lookup the font) silently renders a different font instead.  I *guess*(?) a strategy may be to use something like https://developer.mozilla.org/en-US/docs/Web/CSS/@font-face to explicitly specify a local font path instead.

ambi7 commented:
> As noted in #18039 I don't think this implementation is satisfactory; for example STIXGeneral is only provided by matplotlib on my computer and therefore trying to display in with this mechanism (which uses the browser's OS integration to lookup the font) silently renders a different font instead. I _guess_(?) a strategy may be to use something like https://developer.mozilla.org/en-US/docs/Web/CSS/@font-face to explicitly specify a local font path instead.

@anntzer Thank you for the review! I agree, the prior implementation will not render the correct font. I updated the PR to load the font using @font-face and updated tests. Can you take another look?

anntzer commented:
Does this actually render the right font e.g. from a jupyter notebook?  From local testing it doesn't seem so?

tacaswell commented:
I think it would be better to render to make a temporary `Figure` and put a base64 encodeding png (or a svg with the paths embeded) in the html repr (and also add `_repr_png_`).

tacaswell commented:
@ambi7 Thank you for working on this!  I hope you are not discouraged by the initial feedback.

ambi7 commented:
> Does this actually render the right font e.g. from a jupyter notebook? From local testing it doesn't seem so?

Yes, font-face was having issues loading fonts from file when rendering from a jupyter notebook. I took @tacaswell suggestion and rendered the font via a png image. Looks like it is now rendering the right font by using a temporary `Figure`

anntzer commented:
I like png as well here, but @timhoffm didn't (https://github.com/matplotlib/matplotlib/issues/18039#issuecomment-846591716).

ambi7 commented:
@timhoffm Any additional comments in regards to this implementation?

timhoffm commented:
I can see the argument for images, that decouple the font rendering from the display.

So, this seems the defensive and sane approach. Still I'd like to see HiDPI-support for the rendered images, but that doesn't have to be necessarily part of this PR.

ambi7 reviewed (commented):
resolved by creating _repr_png_ using Figure() to render custom font.

tacaswell commented on lib/matplotlib/font_manager.py:
```diff
@@ -380,6 +382,22 @@ def findSystemFonts(fontpaths=None, fontext='ttf'):
+    font_path = Path(fontent.fname) if fontent.fname != '' else None
+    fig.text(0, 0, fontent.name, font=font_path)
+    with BytesIO() as buf:
+        fig.savefig(buf, bbox_inches='tight', transparent=True)
```
Is this enough to get high-dpi or does this just make things bigger?

```suggestion
        fig.savefig(buf, bbox_inches='tight', transparent=True, dpi=200)
```

QuLogic replied:
That will just make it bigger. The HTML needs to specify that the image is 2x, and I don't know how to tell that to Jupyter via the `_repr_png_` interface.

tacaswell commented:
Cleared @anntzer 's review because it was of the "ask the browser to render it" implementation.

timhoffm commented on lib/matplotlib/tests/test_font_manager.py:
```diff
@@ -268,6 +270,27 @@ def test_fontcache_thread_safe():
+def test_fontentry_dataclass():
+    fontent = FontEntry(name='font-name')
+
+    assert type(fontent.__doc__) == str
```
Do we need to assert this?

ambi7 replied:
No need, removed it

timhoffm commented on lib/matplotlib/tests/test_font_manager.py:
```diff
@@ -268,6 +270,27 @@ def test_fontcache_thread_safe():
+
+    assert type(fontent.__doc__) == str
+
+    png = fontent._repr_png_()
+    html = fontent._repr_html_()
```
Move to the respective places of usage. That's simpler to understand than defining stuff up front and only use it later.

ambi7 replied:
Done!

timhoffm commented:
Thanks @ambi7!
